webapp/channels/src/components/admin_console/types.ts

```diff
@@ -18,188 +22,168 @@ export type ConsoleAccess = {read: {[key: string]: boolean}; write: {[key: strin
-    help_text?: string | JSX.Element;
-    help_text_default?: string | JSX.Element;
-    help_text_markdown?: boolean;
-    help_text_values?: {[key: string]: any};
-    disabled_help_text?: string;
-    disabled_help_text_default?: string;
-    disabled_help_text_markdown?: boolean;
     options: AdminDefinitionSettingDropdownOption[];
     isHelpHidden?: Check;
 }
 
 type AdminDefinitionSettingFileUpload = AdminDefinitionSettingBase & {
     type: 'fileupload';
-    help_text: string;
-    help_text_default: string;
-    remove_help_text: string;
-    remove_help_text_default: string;
-    remove_button_text: string;
-    remove_button_text_default: string;
-    removing_text: string;
-    removing_text_default: string;
-    uploading_text: string;
-    uploading_text_default: string;
+    remove_help_text: string | MessageDescriptor;
+    remove_button_text: string | MessageDescriptor;
+    removing_text: string | MessageDescriptor;
+    uploading_text: string | MessageDescriptor;
     fileType: string;
-    upload_action: () => void;
+    upload_action: (file: File, success: (data: any) => void, error: (err: any) => void) => void;
     set_action?: () => void;
     setFromMetadataField?: string;
-    remove_action: () => void;
+    remove_action: (success: (data: any) => void, error: (err: any) => void) => void;
 }
 
 type AdminDefinitionSettingJobsTable = AdminDefinitionSettingBase & {
     type: 'jobstable';
-    job_type: string;
-    help_text: string;
-    help_text_markdown: boolean;
-    help_text_default: string;
-    help_text_values?: {[key: string]: any};
+    job_type: JobType;
     render_job: Component;
 };
 
 type AdminDefinitionSettingLanguage = AdminDefinitionSettingBase & {
     type: 'language';
-    help_text?: string;
-    help_text_markdown?: boolean;
-    help_text_default?: string;
-    help_text_values?: {[key: string]: any};
     multiple?: boolean;
-    no_result?: string;
-    no_result_default?: string;
-    not_present?: string;
-    not_present_default?: string;
+    no_result?: string | MessageDescriptor;
 }
 
 type AdminDefinitionSettingButton = AdminDefinitionSettingBase & {
     type: 'button';
-    action: () => void;
-    loading?: string;
-    loading_default?: string;
-    help_text?: string | JSX.Element;
-    help_text_default?: string | JSX.Element;
-    help_text_markdown?: boolean;
-    help_text_values?: {[key: string]: any};
-    error_message: string;
-    error_message_default: string;
-    success_message?: string;
-    success_message_default?: string;
+    action: (success: (data?: any) => void, error: (error: {message: string; detailed_error?: string}) => void, siteUrl: string) => void;
+    loading?: string | MessageDescriptor;
+    error_message: string | MessageDescriptor;
+    success_message?: string | MessageDescriptor;
     sourceUrlKey?: string;
 }
 
+type AdminDefinitionSettingUsername = AdminDefinitionSettingBase & {
+    type: typeof Constants.SettingsTypes.TYPE_USERNAME;
```

Comment:
The magic is in `utils/constants.tsx`. If you do something like `const foo = 'somestring' as const;`, `foo` will be of type `'somestring'` instead of type `string`.

Reply:
Ah, I missed that you changed the constants. It still seems like we might want to use an enum for that, but that's outside the scope of this PR which is already big enough